// @nangohq/runner-sdk/dist/errors.js
export class SDKError extends Error {
    payload;
    constructor(payload) {
        super();
        this.payload = payload || {};
    }
}
export class AbortedSDKError extends SDKError {
    code = 'script_aborted';
}
export class UnknownProviderSDKError extends SDKError {
    code = 'unknown_provider_template_in_config';
}
export class InvalidRecordSDKError extends SDKError {
    code = 'invalid_sync_record';
}
export class InvalidActionInputSDKError extends SDKError {
    code = 'invalid_action_input';
}
export class InvalidActionOutputSDKError extends SDKError {
    code = 'invalid_action_output';
}
/**
 * For external use only
 */
export class ActionError extends Error {
    type;
    payload;
    constructor(payload) {
        super();
        this.type = 'action_script_runtime_error';
        if (payload) {
            this.payload = payload;
        }
    }
}
export class ExecutionError extends Error {
    type;
    payload;
    status;
    additional_properties;
    telemetryBag;
    constructor(payload) {
        super();
        this.type = payload.type;
        this.payload = payload.payload;
        this.status = payload.status;
        this.additional_properties = payload.additional_properties;
        this.telemetryBag = payload.telemetryBag;
    }
    toJSON() {
        return {
            type: this.type,
            payload: this.payload,
            status: this.status,
            additional_properties: this.additional_properties
        };
    }
}
