// @nangohq/runner-sdk/dist/sync.js
import { NangoActionBase } from './action.js';
import { validateData } from './dataValidation.js';
export const BASE_VARIANT = 'base';
export class NangoSyncBase extends NangoActionBase {
    variant = BASE_VARIANT;
    lastSyncDate;
    track_deletes = false;
    constructor(config) {
        super(config);
        if (config.lastSyncDate) {
            this.lastSyncDate = config.lastSyncDate;
        }
        if (config.track_deletes) {
            this.track_deletes = config.track_deletes;
        }
        if (config.syncVariant) {
            this.variant = config.syncVariant;
        }
    }
    modelFullName(model) {
        if (this.variant === BASE_VARIANT) {
            return model;
        }
        return `${model}::${this.variant}`;
    }
    /**
     * @deprecated please use batchSave
     */
    async batchSend(results, model) {
        return this.batchSave(results, model);
    }
    validateRecords(model, records) {
        // Validate records
        const hasErrors = [];
        for (const record of records) {
            const validation = validateData({
                version: this.syncConfig?.version || '1',
                input: JSON.parse(JSON.stringify(record)),
                jsonSchema: this.syncConfig.models_json_schema,
                modelName: model
            });
            if (validation === true) {
                continue;
            }
            hasErrors.push({ data: record, validation });
            if (this.runnerFlags?.validateSyncRecords) {
                break;
            }
        }
        return hasErrors;
    }
    removeMetadata(results) {
        if (!Array.isArray(results)) {
            return results;
        }
        return results.map((result) => {
            if (result && typeof result === 'object' && '_nango_metadata' in result) {
                delete result._nango_metadata;
            }
            return result;
        });
    }
}
