// @nangohq/providers/dist/localization.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
const __filename = fileURLToPath(import.meta.url);
const pkgRoot = path.join(__filename, '../../');
let overridesCache = {};
export function clearLocalizationCache() {
    overridesCache = {};
}
export function getLocalizedProviders(baseProviders, language) {
    const languageOverrides = getLanguageOverrides(language);
    return applyLanguageOverrides(baseProviders, languageOverrides);
}
function getLanguageOverrides(language) {
    if (overridesCache[language]) {
        return overridesCache[language];
    }
    const overrides = loadLanguageOverrides(language);
    overridesCache[language] = overrides;
    return overrides;
}
export function loadLanguageOverrides(language) {
    try {
        const languageFilePath = path.join(pkgRoot, 'i18n', `providers.${language}.json`);
        if (!fs.existsSync(languageFilePath)) {
            return {};
        }
        const fileContent = fs.readFileSync(languageFilePath, 'utf8');
        const overrides = JSON.parse(fileContent);
        return overrides || {};
    }
    catch (err) {
        console.error(`Failed to load language overrides for ${language}`, err);
        return {};
    }
}
function applyLanguageOverrides(baseProviders, overrides) {
    return deepMerge(baseProviders, overrides);
}
function deepMerge(target, source) {
    const result = { ...target };
    for (const key in source) {
        if (source[key] && typeof source[key] === 'object' && !Array.isArray(source[key])) {
            result[key] = deepMerge(result[key] || {}, source[key]);
        }
        else {
            result[key] = source[key];
        }
    }
    return result;
}
